### PR TITLE
spirv-val: Add Vulkan decoration interface

### PR DESCRIPTION
**Pull request by sjfricke**
the successor of https://github.com/KhronosGroup/SPIRV-Tools/pull/4293/ and closes https://github.com/KhronosGroup/SPIRV-Tools/issues/3154 and https://github.com/KhronosGroup/SPIRV-Tools/issues/3066

This adds the following:
- VUID-StandaloneSpirv-Flat-04744
- VUID-StandaloneSpirv-Flat-06201
- VUID-StandaloneSpirv-Flat-06202

### CODE REVIEW

**Review by alan-baker — APPROVED** [review]

**Comment by y-novikov** [comment]
This triggers in angle_unittests when I try to roll into Chromium
https://chromium-review.googlesource.com/c/chromium/src/+/3752850
@alan-baker, could you PTAL where the problem is?

**Comment by ShabbyX** [comment]
This is due to an ANGLE bug, we didn't emit `DecorationFlat` for `gl_SampleID`

File: source/val/validate_decorations.cpp [hunk 1]
```diff
@@ -811,6 +811,56 @@ spv_result_t CheckDecorationsOfEntryPoints(ValidationState_t& vstate) {
...
+                       << entry_point << ".";
+              } else if (storage_class == SpvStorageClassOutput &&
+                         (models->size() > 1 || has_frag)) {
+                return vstate.diag(SPV_ERROR_INVALID_ID, var_instr)
+                       << vstate.VkErrorID(6201)
+                       << "OpEntryPoint interfaces variable must not be "
+                          "fragment "
+                          "execution model with an output storage class for "
+                          "Entry Point id "
+                       << entry_point << ".";
```
Comment (eddyb) [inline source/val/validate_decorations.cpp:844]: I don't understand the phrasing used here:
* "must not be {vertex,fragment} execution model"
  * how can a variable *be an execution model*?
  (is there a missing "for" in the middle there?)
* it (effectively) claims vertex shaders can't have inputs and fragment shaders can't have outputs?!
  * without looking up e.g. [`VUID-StandaloneSpirv-Flat-06201` in the spec](https://registry.khronos.org/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-StandaloneSpirv-Flat-06201) there is no way *at all* to tell that the error is talking about *decorations* on interface variables, not their storage classes
  (let alone *which* decorations, i.e. `Flat`/`NoPerspective`/`Sample`/`Centroid`)

It almost looks like there's a place left in the message for printing the decoration name, but that was forgotten? (e.g. after "must not be" one could add "decorated with Flat", though the rest of the message needs a bit of tweaking too)

Comment (eddyb) [reply source/val/validate_decorations.cpp:844]: Looking at the PR this replaces, [its version of the message](https://github.com/KhronosGroup/SPIRV-Tools/pull/4293/files#diff-de97978088736c0ebd7581fb63230936ae4c122bf69f903053ec4c9168315c96) uses `LogStringForDecoration` and overall seems readable, unlike this PR.

Despite the GH username change, it looks like both PRs are by @sjfricke so I don't understand how the message regression happened.

Comment (sjfricke) [reply source/val/validate_decorations.cpp:844]: So I agree the error message is a little vague, can improve that! (unless you were planning on making the PR)

Sorry, I forgot about `LogStringForDecoration` ... I made both these PR a year apart and there were some spec changes in between that time too that was blocking the original PR.

**Pull request merged**